File: 3563.lexicographically-smallest-string-after-adjacent-removals.go
#
# @lc app=leetcode id=3563 lang=golang
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#

# @lc code=start
func lexicographicallySmallestString(s string) string {
	n := len(s)
	if n == 0 {
		return ""
	}
	ss := []byte(s)
	canEmpty := make([][]bool, n)
	for i := range canEmpty {
		canEmpty[i] = make([]bool, n)
	}
	isConsec := func(a, b byte) bool {
		d := int(a) - int(b)
		if d < 0 {
			d = -d
		}
		return d == 1 || d == 25
	}
	for leng := 2; leng <= n; leng += 2 {
		for i := 0; i <= n-leng; i++ {
			j := i + leng - 1
			for k := i; k < j; k++ {
				leftOk := (k == i || canEmpty[i][k-1])
				mStart, mEnd := k+1, j-1
				midOk := (mStart > mEnd || canEmpty[mStart][mEnd])
				if isConsec(ss[k], ss[j]) && leftOk && midOk {
					canEmpty[i][j] = true
					break
				}
			}
		}
	}
	best := make([]string, n+1)
	best[n] = ""
	emptyable := func(l, r int) bool {
		if l > r {
			return true
		}
		if l == r {
			return false
		}
		return canEmpty[l][r]
	}
	for pos := n-1; pos >= 0; pos-- {
		var minS string
		have := false
		if emptyable(pos, n-1) {
			minS = ""
			have = true
		}
		for j := pos; j < n; j++ {
			if emptyable(pos, j-1) {
				cand := string(ss[j]) + best[j+1]
				if !have || cand < minS {
					minS = cand
					have = true
				}
			}
		}
		best[pos] = minS
	}
	return best[0]
}
# @lc code=end

File: 3563.lexicographically-smallest-string-after-adjacent-removals.java
#
# @lc app=leetcode id=3563 lang=java
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#

# @lc code=start
class Solution {
    public String lexicographicallySmallestString(String s) {
        int n = s.length();
        boolean[][] can = new boolean[n+1][n+1];
        for (int i = 0; i <= n; i++) {
            can[i][i] = true;
        }
        for (int len = 1; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len;
                boolean possible = false;
                for (int k = i + 1; k < j; k++) {
                    if (can[i][k] && can[k][j]) {
                        possible = true;
                        break;
                    }
                }
                if (len >= 2) {
                    char left = s.charAt(i);
                    char right = s.charAt(j - 1);
                    int dx = Math.abs((left - 'a') - (right - 'a'));
                    boolean consec = (dx == 1) || (left == 'a' && right == 'z') || (left == 'z' && right == 'a');
                    if (consec && can[i + 1][j - 1]) {
                        possible = true;
                    }
                }
                can[i][j] = possible;
            }
        }
        String[] best = new String[n + 1];
        best[n] = "";
        for (int i = n - 1; i >= 0; i--) {
            String minStr = s.substring(i);
            if (can[i][n]) {
                minStr = "";
            }
            for (int k = i; k < n; k++) {
                if (can[i][k]) {
                    String cand = s.charAt(k) + best[k + 1];
                    if (cand.compareTo(minStr) < 0) {
                        minStr = cand;
                    }
                }
            }
            best[i] = minStr;
        }
        return best[0];
    }
}
# @lc code=end